### PR TITLE
[text_input] prepare for custom text input sources

### PR DESCRIPTION
**Pull request by jpnurmi**
## Description

The existing `TextInputConnection` class has been made abstract and method channel-agnostic. There is now a separate internal class, `_TextInputChannelConnection`, that implements the default type of text input connection that communicates with platform plugins via the text input system channel.

The main motivation of the separation is to allow custom text input sources that can replace the method channel-based implementation and for example, communicate with an in-app virtual keyboard on (embedded) platforms that do not have a system-provided keyboard.

To be able to eliminate direct method channel calls from the `TextInput` class, the following new methods have been introduced:
- `TextInput.detach()`
- `TextInput.reset()`
- `TextInputConnection.hide()`
- `TextInputConnection.setClient()`
- `TextInputConnection.clearClient()`

Consequently, the following methods have been deprecated:
- `TextInputConnection.close()` in favor of `TextInput.detach()`
- `TextInputConnection.connectionClosedReceived()` in favor of `TextInput.reset()`

## Related PRs

This PR was split off from #69146.

## Related Issues

#68988

## Tests

I added the following tests:
- `test/services/text_input_test.dart`:
  - _TextInput message channels_:
    - _text input client is requested to hide on detach_
    - _old client is detached when a new client is attached_
    - _text input connection is reset_

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jpnurmi** [comment]
Rebased again in hopes to pass the CI (unrelated git fetch failures on macOS).

**Review by justinmc — COMMENTED** [review]
Thanks for splitting this out from https://github.com/flutter/flutter/pull/69146, should be easier to review and merge.

It looks like this could hypothetically break someone that was subclassing TextInputConnection.  Do we know if anyone might be doing that?  I've submitted this PR to have the internal Google tests run on it, and if it passes there then I'm more confident that it's not a problem.

CC @LongCatIsLooong since you reviewed https://github.com/flutter/flutter/pull/69146.

**Comment by jpnurmi** [comment]
Thanks for the review!

> It looks like this could hypothetically break someone that was subclassing TextInputConnection. Do we know if anyone might be doing that? I've submitted this PR to have the internal Google tests run on it, and if it passes there then I'm more confident that it's not a problem.

I believe it's not a breaking change because the constructor was private before, so it wasn't possible to create a `TextInputConnection` instance from the outside of `text_input.dart`.

**Comment by justinmc** [comment]
Ah I missed that, you're right!  Should be fine then.  I'm on board once all the tests pass.

File: packages/flutter/lib/src/services/text_input.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_input.dart:973]: nit: keep the old documentation, the deprecation notice should be enough for saying this is replaced by `detach`.

File: packages/flutter/lib/src/services/text_input.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_input.dart:1291]: any chance we can set the handler when we attach to a new connection and unset when we detach from this connection? It sounds wrong that when we switch connection the old handler is still listening to the text input channel

Comment (jpnurmi) [reply packages/flutter/lib/src/services/text_input.dart:1291]: I recall we had it like that in the beginning, but there was a problem with how `_DefaultBinaryMessenger` deal with unhandled messages: https://github.com/flutter/flutter/pull/69146#discussion_r523240713

**Review by justinmc — APPROVED** [review]
LGTM but I'll wait for @LongCatIsLooong's final approval.

File: packages/flutter/lib/src/services/text_input.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_input.dart:1296]: Ah right thanks for pointing that out. It feels like we should let `TextInputConnection` own the text input channel (at least pretend to own it since it's a singleton), and decide what to do when it disconnects? Now that `TextInputConnection` handles the communication between the client and the keyboard,  should `TextInput` just be responsible for creating/switching connections (attach, detach)? It's OK to leave `finishAutofillContext` in `TextInput` for now since we can still access the method channel via `SystemChannels.textInput`. WDYT?

Comment (jpnurmi) [reply packages/flutter/lib/src/services/text_input.dart:1296]: We are still missing a good place for static `TextInputConnection` init and cleanup. The upcoming `TextInputSource` could help with that because it could have such methods that init and cleanup things that apply to all specific type of input connection instances.

The default `TextInputSource` implementation could act as the "owner" of the input channel and pass it over to input connections. It could also set a stub method call handler that discards all incoming messages when not in use.

> It's OK to leave finishAutofillContext in TextInput for now since we can still access the method channel via SystemChannels.textInput.

Accessing `SystemChannels.textInput` directly would unfortunately break autofill tests that rely on `TextInput.setChannel()`. We could actually solve this problem too with `TextInputSource`, though. Since `finishAutofillContext()` is not tied to any specific input connection, we could redirect it to the current input source.

Comment (jpnurmi) [reply packages/flutter/lib/src/services/text_input.dart]: I added an internal `_TextInputSourrce` to address these issues. I'll follow up with the rest in #69146.

File: packages/flutter/lib/src/services/text_input.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_input.dart:1503]: should we call clean up on the previous channel?

File: packages/flutter/lib/src/services/text_input.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_input.dart:1387]: This class will be made public in a different PR?

File: packages/flutter/lib/src/services/text_input.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_input.dart:1440]: nit: maybe crossreference `connectionClosed` in the documentation.

**Review by LongCatIsLooong — APPROVED** [review]
LGTM. Thank you!

File: packages/flutter/lib/src/services/text_input.dart
Comment (jpnurmi) [reply packages/flutter/lib/src/services/text_input.dart:1387]: The next step is to make this an abstract public class and move the default implementation into a private subclass. We have most of it already covered in the original PR #69146 but especially the docs need some work. I will rebase #69146 and continue where we left off.

File: packages/flutter/lib/src/services/text_input.dart
Comment (jpnurmi) [reply packages/flutter/lib/src/services/text_input.dart:1508]: This is only used in tests for setting up a testable environment via `@visibleForTesting TextInput.setChannel()`. This will be moved to the upcoming private default implementation of `TextInputSource` so it won't be visible to the outside.

**Comment by jpnurmi** [comment]
Is the Google testing failure somehow related to this?

**Comment by LongCatIsLooong** [comment]
Doesn't seem to be the case. I have restarted the test.

**Comment by LongCatIsLooong** [comment]
The failure seems to be caused by something irrelevant. Could you rebase the branch to the latest commit and see if the failure goes away?

**Comment by jpnurmi** [comment]
Unrelated Mac build_tests failure. Let's try again.

**Pull request merged**
